### PR TITLE
Explicit outer loops configuration

### PR DESCRIPTION
**Pull request by geofjamg**
**Please check if the PR fulfills these requirements**

- [x] The commit message follows our guidelines
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem?**

No

**What kind of change does this PR introduce?**

Feature

**What is the current behavior?**

We cannot change the outer loop configuration (change the order for instance)

**What is the new behavior (if this is a feature change)?**
We can thanks to a parameter specify explicitly the list of outer loops to run. 

**Does this PR introduce a breaking change or deprecate an API?**

- [ ] The *Breaking Change* or *Deprecated* label has been added
- [ ] The migration guide has been updated in the github wiki *(What changes might users need to make in their application due to this PR?)*

**Other information**:

### CODE REVIEW

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java [hunk 1]
```diff
@@ -1201,6 +1221,13 @@ private static NewtonRaphsonStoppingCriteria createNewtonRaphsonStoppingCriteria
         }
     }
 
+    static List<AcOuterLoop> createOuterLoops(LoadFlowParameters parameters, OpenLoadFlowParameters parametersExt) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java:1224]: So, just to be totally clear:
- If we have the parameter `outerLoopNames`, this override totally the parameters that defines the mode (phase control mode, transformer voltage control mode, shunt voltage control mode).
-  But it seems that it overrides also the booleans that defines if controls are enabled or not. Is it really what you want ? I am not sure we will change the list each time we want to disable a control (and I do it all the time).

**Review by annetill — COMMENTED** [review]
Few questions.

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java (hunk 1 above)
Comment (Hadrien-Godard) [reply src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java:1224]: I agree with Anne.

File: src/main/java/com/powsybl/openloadflow/AbstractAcOuterLoopConfig.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        return new DistributedSlackOuterLoop(activePowerDistribution, parametersExt.isThrowsExceptionInCaseOfSlackDistributionFailure(), parametersExt.getSlackBusPMaxMismatch());
+    }
+
+    protected static IncrementalTransformerVoltageControlOuterLoop createIncrementalTransformerVoltageControlOuterLoop(OpenLoadFlowParameters parametersExt) {
+        return new IncrementalTransformerVoltageControlOuterLoop(parametersExt.getIncrementalTransformerVoltageControlOuterLoopMaxTapShift());
+    }
+
+    protected static ReactiveLimitsOuterLoop createReactiveLimitsOuterLoop(OpenLoadFlowParameters parametersExt) {
+        final double effectiveMaxReactivePowerMismatch;
+        switch (parametersExt.getNewtonRaphsonStoppingCriteriaType()) {
```
[the remark this answers is not in the document]
Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/AbstractAcOuterLoopConfig.java]: Done

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java (hunk 1 above)
Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java:1224]: Fixed, now control status is also used in explicit outer loops configuration

**Review by Hadrien-Godard — APPROVED** [review]

File: src/test/resources/saReport.txt [hunk 3]
```diff
@@ -6,27 +6,27 @@
          Network balance: active generation=607.0 MW, active load=600.0 MW, reactive generation=0.0 MVar, reactive load=200.0 MVar
       + Pre-contingency simulation
          AC load flow complete with NR status 'CONVERGED'
-        + Outer loop Distributed slack on generation
```
Comment (annetill) [inline src/test/resources/saReport.txt:9]: You don't have `on generation` any more.

File: src/test/java/com/powsybl/openloadflow/OpenLoadFlowParametersTest.java [hunk 4]
```diff
@@ -316,7 +320,31 @@ void testCloneParameters() {
...
     }
+
+    @Test
+    void testExplicitOuterLoopsParameter() {
+        LoadFlowParameters parameters = new LoadFlowParameters()
+                .setShuntCompensatorVoltageControlOn(true);
+        OpenLoadFlowParameters parametersExt = new OpenLoadFlowParameters()
+                .setSecondaryVoltageControl(true);
+
+        assertEquals(List.of("DistributedSlack", "VoltageMonitoring", "ReactiveLimits", "ShuntVoltageControl", "SecondaryVoltageControl"), OpenLoadFlowParameters.createOuterLoops(parameters, parametersExt).stream().map(OuterLoop::getType).toList());
```
Comment (annetill) [inline src/test/java/com/powsybl/openloadflow/OpenLoadFlowParametersTest.java:334]: Can we imagine an enum for the different outer loop type to ease setting?

File: src/main/java/com/powsybl/openloadflow/AbstractAcOuterLoopConfig.java [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+            return Optional.of(outerLoop);
+        }
+        return Optional.empty();
+    }
+
+    protected static Optional<AcOuterLoop> createShuntVoltageControlOuterLoop(LoadFlowParameters parameters, OpenLoadFlowParameters parametersExt) {
+        return createShuntVoltageControlOuterLoop(parameters, parametersExt.getShuntVoltageControlMode());
+    }
+
+    protected static Optional<AcOuterLoop> createPhaseShifterControlOuterLoop(LoadFlowParameters parameters, OpenLoadFlowParameters.PhaseShifterControlMode controlMode) {
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/AbstractAcOuterLoopConfig.java:93]: In general, I don't like the use of phase shifter inside open loadflow as we have chosen phase control.

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java [hunk 6]
```diff
@@ -244,7 +248,8 @@ private static <E extends Enum<E>> List<Object> getEnumPossibleValues(Class<E> e
...
         new Parameter(ASYMMETRICAL_PARAM_NAME, ParameterType.BOOLEAN, "Asymmetrical calculation", LfNetworkParameters.ASYMMETRICAL_DEFAULT_VALUE),
         new Parameter(MIN_NOMINAL_VOLTAGE_TARGET_VOLTAGE_CHECK_PARAM_NAME, ParameterType.DOUBLE, "Min nominal voltage for target voltage check", LfNetworkParameters.MIN_NOMINAL_VOLTAGE_TARGET_VOLTAGE_CHECK_DEFAULT_VALUE),
-        new Parameter(REACTIVE_POWER_DISPATCH_MODE_PARAM_NAME, ParameterType.STRING, "Generators reactive power from bus dispatch mode", REACTIVE_POWER_DISPATCH_MODE_DEFAULT_VALUE.name(), getEnumPossibleValues(ReactivePowerDispatchMode.class))
+        new Parameter(REACTIVE_POWER_DISPATCH_MODE_PARAM_NAME, ParameterType.STRING, "Generators reactive power from bus dispatch mode", REACTIVE_POWER_DISPATCH_MODE_DEFAULT_VALUE.name(), getEnumPossibleValues(ReactivePowerDispatchMode.class)),
+        new Parameter(OUTER_LOOP_NAMES_PARAM_NAME, ParameterType.STRING_LIST, "Ordered explicit list of outer loop names to run among " + String.join(", ", ExplicitAcOuterLoopConfig.NAMES), OUTER_LOOP_NAMES_DEFAULT_VALUE)
```
Comment (annetill) [inline src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java:252]: run among?

File: src/test/resources/saReport.txt (hunk 3 above)
Comment (geofjamg) [reply src/test/resources/saReport.txt:9]: yes this is on purpose

File: src/main/java/com/powsybl/openloadflow/AbstractAcOuterLoopConfig.java (hunk 5 above)
Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/AbstractAcOuterLoopConfig.java]: Changed

File: src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java (hunk 6 above)
Comment (geofjamg) [reply src/main/java/com/powsybl/openloadflow/OpenLoadFlowParameters.java]: Replaced by 
Ordered explicit list of outer loop names, supported outer loops are IncrementalPhaseControl, ...

**Review by annetill — APPROVED** [review]

**Pull request merged**